subsys/net/lib/azure_fota/azure_fota.c

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <net/fota_download.h>
+#include <net/azure_iot_hub.h>
+#include <net/azure_fota.h>
+#include <cJSON.h>
+#include <cJSON_os.h>
+
+#include "azure_iot_hub_dps.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(azure_fota, CONFIG_AZURE_FOTA_LOG_LEVEL);
+
+/* Enum to keep the fota status */
+enum fota_status {
+	STATUS_NONE = 0,
+	STATUS_DOWNLOADING,
+	STATUS_DOWNLOAD_DONE,
+	STATUS_DOWNLOAD_ERROR,
+	STATUS_APPLY_UPDATE,
+	STATUS_REBOOT,
+};
+
+/* Static variables */
+static azure_fota_callback_t callback;
+static size_t download_progress;
+static enum fota_status fota_state;
+static char hostname[CONFIG_AZURE_FOTA_HOSTNAME_MAX_LEN];
+static char file_path[CONFIG_AZURE_FOTA_FILE_PATH_MAX_LEN];
+
+/* Static functions */
+
+static void fota_dl_handler(const struct fota_download_evt *fota_evt)
+{
+	struct azure_fota_event evt;
+
+	switch (fota_evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		LOG_INF("FOTA download completed evt received");
+
+		fota_state = STATUS_DOWNLOAD_DONE;
+
+		evt.type = AZURE_FOTA_EVT_DL_PROGRESS;
+		evt.dl.progress = AZURE_FOTA_EVT_DL_COMPLETE_VAL;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		evt.type = AZURE_FOTA_EVT_ERASE_PENDING;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		evt.type = AZURE_FOTA_EVT_ERASE_DONE;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		LOG_ERR("FOTA download failed, report back");
+		fota_state = STATUS_NONE;
+		evt.type = AZURE_FOTA_EVT_ERROR;
+		callback(&evt);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		/* Only if CONFIG_FOTA_DOWNLOAD_PROGRESS_EVT is enabled */
+		download_progress = fota_evt->progress;
+		evt.type = AZURE_FOTA_EVT_DL_PROGRESS;
+		evt.dl.progress = download_progress;
+		callback(&evt);
+		break;
+	default:
+		LOG_WRN("Unrecognized Azure FOTA event type: %d",
+			fota_evt->id);
+		break;
+	}
+}
+
+/* @brief Extracts firmware image information from device twin JSON document.
+ *
+ * @retval 0 if successfully parsed, hostname and file_path populated.
+ * @retval -ENOMSG Could not parse incoming buffer as JSON.
+ * @retval -ENOMEM Could not find object, or insufficient memory to allocate it.
+ */
+static int extract_fw_details(const char *msg)
+{
+	int err = -ENOMEM;
+	struct cJSON *root_obj, *desired_obj, *fw_obj, *fw_version_obj,
+		*fw_location_obj, *fw_check_obj, *fw_hostname_obj, *fw_path_obj;
+	uint32_t fw_version, fw_check;
+
+	/* The properties.desired.firmware object has the excpected structure:
+	 *	properties.desired.firmware : {
+	 *		fwVersion : <version>,
+	 *		fwLocation : {
+	 *			host : <hostname>,
+	 *			path : <file path>
+	 *		},
```

Comment:
Is this still needed? The pre-parsing was introduced so that the device could download from AWS S3 using a non-secure URL (it would construct the non-https url using the properties in the struct).

See https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/net/aws_fota.html#limitations

And either way, I would prefer to have a string value here, because in both case it's the users responsibility to provide a non-secure download location.

Reply:
`fwPackageURI` is the property used in [the reference](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update#update-the-firmware)

Reply:
Do you mean why it's split into `host`and `path`? It's just convenience, as that's the format the FOTA download client in NCS requires to have it (it can download multiple images from the same host and apply it to different partitions). I decided to let the cloud side take the job of splitting into host and path instead of parsing it on the device.

Reply:
Alright.

Reply:
> I can change the name into fwPackageURI, although host and path would probably also have to change names from what I read here.

Or maybe not, I read the example wrong.

Reply:
Let's keep it `fwLocation` because `fwPackageURI` in the Azure tutorial is a string, so I'd rather not create an overlap of different types.

Also, it's not a full URI, because `scheme` is omitted, because it's hardcoded to `http`.

Reply:
In Bifravst I have implemented it so that the download URL is stored as string in `fwPackageURI` and parsed in `fwLocation`: https://github.com/bifravst/azure/blob/79a30865eef9c2bbf7c4b301171384a1b5ddb6f8/features/FOTA.feature#L43-L69